### ISSUE TITLE
about batchsize

### ISSUE THREAD

**Issue by leoxxxxxD** [ultralytics/yolov5#635]
## ❔Question
Before the 2.0, I used batch size = 40 for training, now I can only use batch size = 16 for training, 4 * 2080ti. I want to know why？

**Comment by leoxxxxxD** [comment]
@glenn-jocher i am using DataParallel

**Comment by lorenzomammana** [comment]
I'm not using data parallel but after updating to pytorch v1.6 i can't handle 16 images per batch as i was doing before without any problems, didn't have any issue with v2.0 i'm sure it's after the pytorch update that the problem has started.
Are you also using the latest commits?

**Comment by glenn-jocher** [comment]
@Ataraxiaaa see multi gpu training tutorial: https://docs.ultralytics.com/yolov5

**Comment by leoxxxxxD** [comment]
@lorenzomammana all are up to date,i will try pytorch=1.5.1

**Comment by leoxxxxxD** [comment]
@glenn-jocher  I did an experiment on v2.0, when I use pytorch = 1.5.1, I can set batchsize = 40, imgsize = 736, but when I use pytorch = 1.6.0, I can only set batchsize = 16, imgsize = 736,

**Comment by leoxxxxxD** [comment]
@lorenzomammana yeah, you are right

**Comment by lorenzomammana** [comment]
@glenn-jocher I think this is a major problem, probably not related to yolov5 itself but still pretty huge.
Haven't you encounter any issue when training with the latest commits?

**Comment by glenn-jocher** [comment]
@lorenzomammana all aspects of training, testing and inference are working correctly in current origin/master.

**Comment by glenn-jocher** [comment]
@lorenzomammana @Ataraxiaaa FYI make sure if you are using multi-GPU you are operating in DDP mode with torch.distributed.launch, and not using the older syntax, which works but with reduced performance. See the multi-gpu training tutorial for details:
https://docs.ultralytics.com/yolov5

**Comment by lorenzomammana** [comment]
@glenn-jocher I'm having this issue with a single device, rolling back to the 1.5 version of pytorch and everything is working fine.
I was thinking of a problem with the new translate: 0.5 but it's working.

The error is the data loader going out of shared memory after something like 10-15 epochs in 1.6, same training parameters but with 1.5 (+ apex) no errors occours.

The only thing i'm doing differently from your yolo model training is using the multi-scale flage, maybe this is causing the issue.
@Ataraxiaaa are you also using --multi-scale?

**Comment by glenn-jocher** [comment]
@lorenzomammana I've observed slight speed and memory improvements in the recent code (both with single and milti gpu), so the opposite of your observations. It could be your environment, you might want to try using one below. I'm using GCP and Google Colab with the latest origin/master clone.

## Environment

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab Notebook** with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- **Kaggle Notebook** with free GPU: [https://www.kaggle.com/models/ultralytics/yolov5](https://www.kaggle.com/models/ultralytics/yolov5)
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) 
- **Docker Image** https://hub.docker.com/r/ultralytics/yolov5. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) ![Docker Pulls](https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker)

**Comment by lorenzomammana** [comment]
@glenn-jocher I've also observed an incredible speed improvement actually, it takes 7 min on 1.5 for an epoch and 5 min on 1.6, but it's quite useless if the model crashes after a few epochs.
It doesn't look like it's using more gpu memory rather that after some time the shared memory saturates and the program stops.
Unfortunately i'm using a machine without the rights to change the shared memory...

**Comment by glenn-jocher** [comment]
@lorenzomammana 

it appears you have environment problems. In order to run YOLOv5 correctly your environment must meet  the minimum version requirements for the dependencies described in https://github.com/ultralytics/yolov5#requirements. You can either update your local environment to bring it into compliance or you can use one of our verified environment options below.

## Reproduce Our Environment

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab Notebook** with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- **Kaggle Notebook** with free GPU: [https://www.kaggle.com/models/ultralytics/yolov5](https://www.kaggle.com/models/ultralytics/yolov5)
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) 
- **Docker Image** https://hub.docker.com/r/ultralytics/yolov5. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) ![Docker Pulls](https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker)

**Comment by lorenzomammana** [comment]
I'll try to double check my environment then!

**Comment by lorenzomammana** [comment]
I've double check and all the packages are now updated to the latest version, still the error persists.
I'm quite sure it's this problem, https://github.com/pytorch/pytorch/issues/40911, so there is definitely a memory leak in the 1.6 version of pytorch when using amp in certain environments.

**Comment by watertianyi** [comment]
@lorenzomammana @glenn-jocher   
I have the same situation. Maybe there is something wrong with the latest version of Python 1.6.0.
 I use yolov5x.pt as the initialization weight. CFG is yolov5x.yaml, rtx2080ti GPU graphics card is 11g. If you do not use apex and batchsize is set to 8, you will report an error that CUDA video memory is insufficient. If you use apex and batchsize set to 8, you can use it normally. What you used is the latest code yesterday

**Comment by glenn-jocher** [comment]
@honghande you must be using old code, as there is no Apex functionality present anymore. It has been completely removed from the repo.

**Comment by watertianyi** [comment]
@glenn-jocher 
The code is the latest code of yolov5. I deleted the previous apex environment by following the requirements.txt Installation environment, rtx2080gpu memory 11g. Set the batchsize to 8 and retrain. The error is as follows:
![image](https://user-images.githubusercontent.com/29244715/90587377-ce302200-e20b-11ea-9334-128688f24cb1.png)
![image](https://user-images.githubusercontent.com/29244715/90587441-f61f8580-e20b-11ea-926a-3880f52cd2ce.png)
![image](https://user-images.githubusercontent.com/29244715/90587471-07689200-e20c-11ea-9388-80d7867f8478.png)
Two days ago, using batchsize = 8 in rtx2080ti memory as 8g can run, but now it can't, I don't know why?
Before yolov5 V2.0, there is no problem with the same batch size = 8 and rtx2080 GPU with 8g video memory. If you change to the latest yolov5, you will have the above problems;

**Comment by glenn-jocher** [comment]
@honghande CUDA OOM means your GPU is out of RAM. Reduce batch size, img size, or use a GPU with more memory.

**Comment by honghande** [comment]
batchsize=4，the same problem




------------------ Original ------------------

**Comment by watertianyi** [comment]
@glenn-jocher batchsize=4，the same problem

**Comment by glenn-jocher** [comment]
@honghande you may have an environment problem. I would create a new python 3.8 venv and install requirements.txt and go from there. 4 images at 640 sounds too small. 

You can see the batch sizes we use for training on V100's with 16GB of memory here. For v5x we use batch size 16, so 8 should work on your 2080Ti, but I don't actually know since I have not tested on that exact card.
https://github.com/ultralytics/yolov5#training

**Comment by watertianyi** [comment]
@glenn-jocher  I use conda to install the virtual environment. The virtual environment is as follows:
_libgcc_mutex             0.1                        main    defaults
absl-py                   0.9.0                     <pip>
blas                      1.0                         mkl    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free
ca-certificates           2020.6.24                     0    defaults
cachetools                4.1.1                     <pip>
certifi                   2020.6.20                py38_0    defaults
chardet                   3.0.4                     <pip>
cudatoolkit               10.1.243             h6bb024c_0    defaults
cycler                    0.10.0                    <pip>
Cython                    0.29.21                   <pip>
freetype                  2.10.2               h5ab3b9f_0    defaults
google-auth               1.20.1                    <pip>
google-auth-oauthlib      0.4.1                     <pip>
grpcio                    1.31.0                    <pip>
idna                      2.10                      <pip>
intel-openmp              2020.1                      217    defaults
jpeg                      9b                            0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free
kiwisolver                1.2.0                     <pip>
lcms2                     2.11                 h396b838_0    defaults
ld_impl_linux-64          2.33.1               h53a641e_7    defaults
libedit                   3.1.20191231         h14c3975_1    defaults
libffi                    3.3                  he6710b0_2    defaults
libgcc-ng                 9.1.0                hdf63c60_0    defaults
libpng                    1.6.37               hbc83047_0    defaults
libstdcxx-ng              9.1.0                hdf63c60_0    defaults
libtiff                   4.1.0                h2733197_1    defaults
lz4-c                     1.9.2                he6710b0_1    defaults
Markdown                  3.2.2                     <pip>
matplotlib                3.3.1                     <pip>
mkl                       2020.1                      217    defaults
mkl-service               2.3.0            py38he904b0f_0    defaults
mkl_fft                   1.1.0            py38h23d657b_0    defaults
mkl_random                1.1.1            py38h0573a6f_0    defaults
ncurses                   6.2                  he6710b0_1    defaults
ninja                     1.7.2                         0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free
numpy                     1.19.1           py38hbc911f0_0    defaults
numpy-base                1.19.1           py38hfa32c7d_0    defaults
oauthlib                  3.1.0                     <pip>
olefile                   0.46                       py_0    defaults
opencv-python             4.1.2.30                  <pip>
openssl                   1.1.1g               h7b6447c_0    defaults
pillow                    7.2.0            py38hb39fc2d_0    defaults
pip                       20.2.2                   py38_0    defaults
protobuf                  3.13.0                    <pip>
pyasn1                    0.4.8                     <pip>
pyasn1-modules            0.2.8                     <pip>
pyparsing                 2.4.7                     <pip>
python                    3.8.5                hcff3b4d_0    defaults
python-dateutil           2.8.1                     <pip>
pytorch                   1.6.0           py3.8_cuda10.1.243_cudnn7.6.3_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
PyYAML                    5.3.1                     <pip>
readline                  8.0                  h7b6447c_0    defaults
requests                  2.24.0                    <pip>
requests-oauthlib         1.3.0                     <pip>
rsa                       4.6                       <pip>
scipy                     1.5.2                     <pip>
setuptools                49.6.0                   py38_0    defaults
six                       1.15.0                     py_0    defaults
sqlite                    3.32.3               h62c20be_0    defaults
tensorboard               2.3.0                     <pip>
tensorboard-plugin-wit    1.7.0                     <pip>
tk                        8.6.10               hbc83047_0    defaults
torchvision               0.7.0                py38_cu101    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
tqdm                      4.48.2                    <pip>
urllib3                   1.25.10                   <pip>
Werkzeug                  1.0.1                     <pip>
wheel                     0.34.2                   py38_0    defaults
xz                        5.2.5                h7b6447c_0    defaults
zlib                      1.2.11                        0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/pkgs/free
zstd                      1.4.5                h9ceee32_0    defaults

**Comment by watertianyi** [comment]
@glenn-jocher Now I can set batch size = 4, but I can't set -- multi scale. The above problem will appear when setting it? Yolov5x.yaml and yolov5x.pt were used,
![image](https://user-images.githubusercontent.com/29244715/90635839-4f170a00-e25c-11ea-9b7f-833ec79386ad.png)
![image](https://user-images.githubusercontent.com/29244715/90635981-87b6e380-e25c-11ea-8e60-ac6f8a37afdb.png)
Does yolov5x consume memory this way？


If imgsize is set to 416, you can use batchsize = 16,
![image](https://user-images.githubusercontent.com/29244715/90637253-58a17180-e25e-11ea-8079-b8e285823572.png)
![image](https://user-images.githubusercontent.com/29244715/90637312-6a831480-e25e-11ea-9009-3cb310128a77.png)

**Comment by glenn-jocher** [comment]
@honghande --multi-scale uses more memory. I would recommend simply training with the default settings (which don't include --multi-scale, etc.).

**Comment by watertianyi** [comment]
@glenn-jocher 
![image](https://user-images.githubusercontent.com/29244715/90709941-e9646580-e2cf-11ea-9da8-58be10a4cd90.png)
For user-defined data training, single category, with V5 x, why does the recall rate drop so fast? How to analyze and improve the accuracy and recall rate?

**Comment by glenn-jocher** [comment]
@hande6688 recall, precision do not matter, as these are user-settable properties. You can obtain any recall you want, from 0 to 1, by varying the confidence threshold during inference.

What you should be asking yourself is how reduce val loss overfitting.

**Comment by watertianyi** [comment]
@glenn-jocher Thank you for your timely reply. Yes, how can we reduce reduce val loss overfitting.,

**Comment by glenn-jocher** [comment]
@hande6688 I don't know, that's an excellent question! For training from scratch I have an open issue myself on this:
https://github.com/ultralytics/yolov5/issues/746

For finetuning the requirements may differ, but you can get started with any of the same ideas. For finetuning the official YOLOv5 models on VOC 50 epochs, I found increased scale jitter and mixup helped reduce overfitting. Essentially more aggressive augmentation policy seems to benefit finetuning moreso than training from scratch.

In your particular case, your obj loss gain is obviously too high, as your obj loss is 3X larger than your regression loss. You need to balance these if you expect them to deliver balanced contributions to your solutions.
